Implement formgrader as a jupyterhub service

jhamrick opened the pull request:
Fixes #568 

I still need to update the tests and documentation, but I think this is the bulk of the implementation that needs to happen. This will work out of the box as a JupyterHub-managed service, but you can also use it as an externally managed service if you set the right config options/pass the same environment variables that JupyterHub would pass.

As a minimal test case, I have it working with the following `jupyterhub_config.py`:

```python
c = get_config()

# Add users here that are allowed admin access to JupyterHub.
c.Authenticator.admin_users = ["jhamrick"]

# Add users here that are allowed to login to JupyterHub.
c.Authenticator.whitelist = [
    "jhamrick"
]

c.JupyterHub.services = [
    {
        'name': 'formgrader',
        'admin': True,
        'command': ['nbgrader', 'formgrade', '--log-level=DEBUG'],
        'url': 'http://127.0.0.1:9000'
    }
]
```

and `nbgrader_config.py`:

```python
import os

c = get_config()

## Generic nbgrader options (technically, the exchange directory is not used
## by the formgrader, but it is left in here for consistency with the rest of
## the user guide):

c.NbGrader.course_id = "example_course"
c.TransferApp.exchange_directory = "/tmp/exchange"
c.NbGrader.db_assignments = [dict(name="ps1", duedate="2015-02-02 17:00:00 UTC")]
c.NbGrader.db_students = [
    dict(id="bitdiddle", first_name="Ben", last_name="Bitdiddle"),
    dict(id="hacker", first_name="Alyssa", last_name="Hacker"),
    dict(id="reasoner", first_name="Louis", last_name="Reasoner")
]

## Options that are specific to the formgrader and integrating it with JuptyerHub:

c.FormgradeApp.ip = "127.0.0.1"
c.FormgradeApp.port = 9000
c.FormgradeApp.authenticator_class = "nbgrader.auth.hubauth.HubAuth"

# Change this to be the path to the user guide folder in your clone of
# nbgrader, or just wherever you have your class files. This is relative
# to the root of the notebook server launched by JupyterHub, which is
# probably your home directory. This is used for accessing the *live*
# version of notebooks via JupyterHub. If you don't want to access the
# live notebooks and are fine with just the static interface provided by
# the formgrader, then you can ignore this option.
c.HubAuth.notebook_url_prefix = "path/to/class_files"

# Change this to be the list of unix usernames that are allowed to access
# the formgrader.
c.HubAuth.graders = ["jhamrick"]
```

cc @ellisonbg @minrk

ellisonbg commented:
Does this (or can it) use the groups access list on the JupyterHub auth object. Then we could configure the formgrade by just providing the JuptyerHub group that should be able to access it. That would prepare us for later hosting formgrade in shared notebook servers.

jhamrick commented:
Yes, I believe that should be possible...

However, I am finding that access to the live notebook will not work as a Jupyter service, at least not the way it is currently, because the service is not allowed to request admin access to another user's server (or rather, there is a bug that causes this to fail -- but probably that shouldn't be something services can request? @minrk ):

```
[W 2017-01-06 16:34:01.369 JupyterHub users:201] Admin user formgrader has requested access to quux's server
[D 2017-01-06 16:34:01.372 JupyterHub base:236] Setting cookie for quux: jupyter-hub-token-quux, {}
[E 2017-01-06 16:34:01.372 JupyterHub web:1548] Uncaught exception POST /hub/api/users/quux/admin-access (127.0.0.1)
    HTTPServerRequest(protocol='http', host='127.0.0.1:57831', method='POST', uri='/hub/api/users/quux/admin-access', version='HTTP/1.1', remote_ip='127.0.0.1', headers={'Host': '127.0.0.1:57831', 'Content-Length': '0', 'Authorization': 'token c9c6aa141fad4a939d9679fa61c24689', 'User-Agent': 'python-requests/2.12.4', 'Accept-Encoding': 'gzip, deflate', 'Connection': 'keep-alive', 'Accept': '*/*'})
    Traceback (most recent call last):
      File "/Users/jhamrick/miniconda3/envs/nbgrader/lib/python3.5/site-packages/tornado/web.py", line 1467, in _execute
        result = method(*self.path_args, **self.path_kwargs)
      File "/Users/jhamrick/project/tools/jupyterhub/jupyterhub/utils.py", line 109, in decorated
        return method(self, *args, **kwargs)
      File "/Users/jhamrick/project/tools/jupyterhub/jupyterhub/apihandlers/users.py", line 211, in post
        current.other_user_cookies.add(name)
    AttributeError: 'Service' object has no attribute 'other_user_cookies'

[FormgradeApp | WARNING] Failed to gain admin access to user quux's server (response: 500 Internal Server Error)
[E 2017-01-06 16:34:01.387 JupyterHub log:99] {
      "Host": "127.0.0.1:57831",
      "Content-Length": "0",
      "Authorization": "token [secret]",
      "User-Agent": "python-requests/2.12.4",
      "Accept-Encoding": "gzip, deflate",
      "Connection": "keep-alive",
      "Accept": "*/*"
    }
[E 2017-01-06 16:34:01.387 JupyterHub log:100] 500 POST /hub/api/users/quux/admin-access (formgrader@127.0.0.1) 21.66ms
```

The exception is that if there is only one person doing the grading, and the notebooks are located in their account, then access to the live server should still be possible. But if there are multiple people doing the grading, then this will not be the case (at least for now 😞 )

jhamrick commented:
Ok, this now determines access using JupyterHub groups rather than having it be an option in the formgrader itself.

jhamrick commented:
Ok, done with docs and tests as well now. I think this should be good to go.

@ellisonbg it would be great if you could test this with your deployment before I merge.

ellisonbg commented:
Very very cool! The deployment looks very simple now. I will test this out now on my test deployment and provide feedback.

ellisonbg commented on nbgrader/docs/source/configuration/jupyterhub/jupyterhub_config.py:
```diff
@@ -7,3 +7,24 @@
+# JupyterHub, you will need to use the JupyterHub API to remove that user
+# from the formgrader group.
+c.JupyterHub.load_groups = {
+    'formgrader': ['instructor1', 'instructor2']
```
I found it a little confusing to have the group name be formgrade because that name is used elsewhere (the name of the service). In practice I name the groups by the course. So maybe something generic but more descriptive such as `course1`?

ellisonbg commented on nbgrader/docs/source/configuration/jupyterhub/jupyterhub_config.py:
```diff
@@ -7,3 +7,24 @@
+# This is the key piece which defines how JupyterHub should talk to nbgrader
+c.JupyterHub.services = [
+    {
+        'name': 'formgrader',
```
I would name this something that makes it clear you have multiple of these such as `formgrader-course1`

ellisonbg commented on nbgrader/docs/source/configuration/jupyterhub/jupyterhub_config.py:
```diff
@@ -7,3 +7,24 @@
+        'name': 'formgrader',
+        'admin': True,
+        'command': ['nbgrader', 'formgrade'],
+
```
I also need to add the following option:

```
'cwd': '/home/instructor1/nbgrader/course'
```

This is the directory where I have my `nbgrader_config.py` file and which also contains all of the grading directories.

ellisonbg commented on nbgrader/docs/source/configuration/jupyterhub/nbgrader_config.py:
```diff
@@ -36,11 +30,11 @@
 # the formgrader, then you can ignore this option.
 c.HubAuth.notebook_url_prefix = "path/to/class_files"
 
```
In practice I needed to add the following option:

`c.HubAuth.grader_group = 'course1'`

I would add this and emphasize it has to match one of the groups created by JupyterHub

ellisonbg commented on nbgrader/docs/source/configuration/jupyterhub/nbgrader_config.py:
```diff
@@ -36,11 +30,11 @@
 # the formgrader, then you can ignore this option.
 c.HubAuth.notebook_url_prefix = "path/to/class_files"
```
I didn't test the live notebook server stuff...

ellisonbg commented on nbgrader/docs/source/configuration/jupyterhub_config.rst:
```diff
@@ -29,94 +29,83 @@ the same system. By integrating with JupyterHub, nbgrader streamlines the
+
+Running this command should launch JupyterHub, which will then be accessible
+from ``localhost:8000``. It will also launch the formgrader, which you should
+be able to access from ``localhost:8000/services/formgrader``. If you
```
If we change the name of the service in the config file we will need to change this as well.

jhamrick commented:
Awesome, thanks for the feedback. I'll implement these changes tomorrow.

jhamrick commented:
Ok, I have updated stuff based on your comments. I made the `HubAuth.grader_group` option required to be specified, that way people have to customize it based on whatever their class name is. And I updated the relevant documentation.

ellisonbg commented:
This looks great. Seems like the only limitation of this approach is that only one person can access the live notebooks. In the future, if we move formgrade entirely to a serverextension and deploy using a shared notebook server that limitation would be addressed as well. I think this is already a huge improvement and that we should cut a release with this approach. Once the shared notebook server stuff is better vetted, we can think about that approach. Thank you so much for working on this. I will be deploying it on Monday morning ;-)

jhamrick commented:
> Seems like the only limitation of this approach is that only one person can access the live notebooks.

Yes, I definitely agree, and that's something I'd like to try to address ASAP as it's definitely a regression from previous behavior 😞 

> I think this is already a huge improvement and that we should cut a release with this approach.

Yes, I agree. I will be hopefully releasing 0.4.0 in the next week or two (not before Monday unfortunately, but certainly soon!)

> Thank you so much for working on this. I will be deploying it on Monday morning ;-)

Awesome, I hope it goes well!

ellisonbg commented:
Great, I will be running on master of jupyterhub and nbgrader for my course
(monday!). I will configure nbgrader using the new approach and also test
out the shared notebook stuff.

On Sat, Jan 7, 2017 at 12:01 PM, Jessica B. Hamrick <
notifications@github.com> wrote:

> Merged #618 <https://github.com/jupyter/nbgrader/pull/618>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jupyter/nbgrader/pull/618#event-915025160>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABr0HGTwZlNXxWZpVcZCtPWUgNM5Sypks5rP-ElgaJpZM4LdKG5>
> .
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com
